segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,336 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
      *
-     * This method retrieves a few eligible chunks for deletion at a time.
-     * The chunk is deleted only if the metadata for it does not exist or is marked inactive.
-     * If there are any errors then failed chunk is enqueued back up to a max number of attempts.
-     * If suspended or there are no items then it "sleeps" for time specified by configuration.
-     *
-     * @param isBackground True if the caller is background task else False if called explicitly.
-     * @param maxItems     Maximum number of items to delete at a time.
-     * @return CompletableFuture which is completed when garbage is deleted.
-     */
-    CompletableFuture<Boolean> deleteGarbage(boolean isBackground, int maxItems) {
-        log.debug("{}: Iteration {} started.", traceObjectId, iterationId.get());
-        // Sleep if suspended.
-        if (suspended.get() && isBackground) {
-            log.info("{}: deleteGarbage - suspended - sleeping for {}.", traceObjectId, config.getGarbageCollectionDelay());
-            return CompletableFuture.completedFuture(false);
-        }
+     * @param operation    The Callable to execute.
+     * @param <R>       Return type of the operation.
+     * @param keyNames The names of the keys involved in this operation (for sequencing purposes).
+     * @return A CompletableFuture that, when completed, will contain the result of the operation.
+     * If the operation failed, it will contain the cause of the failure.
+     * */
+    private <R> CompletableFuture<R> executeSerialized(Callable<CompletableFuture<R>> operation, String... keyNames) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        return this.taskScheduler.add(Arrays.asList(keyNames), () -> executeExclusive(operation, keyNames));
+    }
 
-        // Find chunks to delete.
-        val chunksToDelete = new ArrayList<GarbageChunkInfo>();
-        int count = 0;
+    /**
+     * Executes the given Callable asynchronously and exclusively.
+     * It returns a CompletableFuture that will be completed with the result.
+     * The operations are not allowed to be concurrent.
+     *
+     * @param operation    The Callable to execute.
+     * @param <R>       Return type of the operation.
+     * @param keyNames The names of the keys involved in this operation (for sequencing purposes).
+     * @return A CompletableFuture that, when completed, will contain the result of the operation.
+     * If the operation failed, it will contain the cause of the failure.
+     * */
+    private <R> CompletableFuture<R> executeExclusive(Callable<CompletableFuture<R>> operation, String... keyNames) {
+        return CompletableFuture.completedFuture(null).thenComposeAsync(v -> {
+            Exceptions.checkNotClosed(this.closed.get(), this);
+            try {
+                return operation.call();
+            } catch (CompletionException e) {
+                throw new CompletionException(Exceptions.unwrap(e));
```

Comment:
You can merge both of these into one. `Exceptions.unwrap` will return its argument if it's not a `CompletionException`.

Reply:
Why do you catch `CompletionException` just to rethrow another `CompletionException` with its inner exception?

This whole try-catch can be rewritten as:

```
 try {
                return operation.call();
} catch(Exception ex){
   throw new CompletionException(ex);
}
```

Everywhere we log or handle exceptions, we invoke `Exceptions.unwrap` - if you look in that method it recursively unwraps `CompletionException` until it finds the real exception inside.